src/functions/New-MockObject.ps1

```diff
@@ -27,11 +63,73 @@ https://pester.dev/docs/usage/mocking
 #>
 
     param (
-        [Parameter(Mandatory = $true)]
+        [Parameter(ParameterSetName = "Type", Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [type]$Type,
+        [Parameter(ParameterSetName = "InputObject", Mandatory)]
         [ValidateNotNullOrEmpty()]
-        [type]$Type
+        $InputObject,
+        [Parameter(ParameterSetName = "Type")]
+        [Parameter(ParameterSetName = "InputObject")]
+        [hashtable]$Properties,
+        [Parameter(ParameterSetName = "Type")]
+        [Parameter(ParameterSetName = "InputObject")]
+        [hashtable]$Methods,
+        [string] $MethodHistoryPrefix = "_"
     )
 
-    [System.Runtime.Serialization.Formatterservices]::GetUninitializedObject($Type)
+    $mock = if ($PSBoundParameters.ContainsKey("InputObject")) {
+        $PSBoundParameters.InputObject
+    }
+    else {
+        [System.Runtime.Serialization.Formatterservices]::GetUninitializedObject($Type)
+    }
+
+    if ($null -ne $Properties) {
+        foreach ($property in $Properties.GetEnumerator()) {
+            if ($mock.PSObject.Properties.Item($property.Key)) {
+                $mock.PSObject.Properties.Remove($property.Key)
+            }
+            $mock.PSObject.Properties.Add([Pester.Factory]::CreateNoteProperty($property.Key, $property.Value))
+            }
+        }
+
+        if ($null -ne $Methods) {
+            foreach ($method in $Methods.GetEnumerator()) {
+                $historyName = "$($MethodHistoryPrefix)$($method.Key)"
+                if ($mock.PSObject.Properties.Item($historyName)) {
+                    $mock.PSObject.Properties.Remove($historyName)
+                }
+                $mock.PSObject.Properties.Add([Pester.Factory]::CreateNoteProperty($historyName, [System.Collections.Generic.List[object]]@()))
+
+                $saveHistoryAndInvokeUserScriptBlock = & {
+                    # this new scriptblock ensures we only copy the variables here
+                    # because closure only copies local variables, the scriptblock execution
+                    # returns a scriptblock that is a closure
+
+                    # save the provided scriptblock as $scriptblock in the closure
+                    $scriptBlock = $method.Value
+                    # save history name as $historyName in the closure
+                    $historyName = $historyName
+                    # save count as reference object so we can easily update the value
+                    $count = @{ Count = 0 }
+
+                    {
+                        # before invoking user scriptblock up the counter by 1 and save args
+                        $this.$historyName.Add([PSCustomObject] @{ Call = ++$count.Count; Arguments = $args })
+
+                        # then splat the args, if user specifies parameters in the scriptblock they
+                        # will get the values by order, same as if they called the script method
+                        & $scriptBlock @args
```

Comment:
Limitation: self-referencing with `$this` is not possible in the user-provided scriptblock.

Reply:
Godd point, added it. I am just not sure if the api I am using is compatible with all powershell versions, the build will tell. Otherwise I need to do it the way we do it in Mock. Or I could also change the approach to not require script block local data.